Checkpoint of Euler spiral based stroke expansion

raphlinus opened the pull request:
This branch contains an implementation of stroke expansion based on Euler spirals. At this point, the CPU shader implementation (in `cpu-shader/flatten.rs`) should be a reasonably solid implementation, with care taken for numerical robustness in particular, but more validation is needed.

In particular, some of the "tricky stroke" cases have a semi-circle taken out near the endpoints. This is the expected rendering from slightly perturbing the co-linear cubic Bézier, then drawing just the parallel curve and cap. However, it is not strongly correct rendering (in the sense of the Nehab 2020 "Converting stroked primitives to filled primitives" paper sense).

The next major step is translation of that logic to WGSL so strokes can be fully evaluated on GPU.

Progress toward #303

armansito commented on src/cpu_shader/euler.rs:
```diff
@@ -0,0 +1,277 @@
+use std::f32::consts::FRAC_PI_4;
+
+#[derive(Debug)]
+pub struct CubicParams {
```
I suggest documenting this parameterization for a cubic and briefly explain how it relates to the Euler spiral, in a few words.

armansito commented on src/cpu_shader/euler.rs:
```diff
@@ -0,0 +1,277 @@
+    pub fn from_points_derivs(p0: Vec2, p1: Vec2, q0: Vec2, q1: Vec2, dt: f32) -> Self {
+        let chord = p1 - p0;
+        // Robustness note: we must protect this function from being called when the
+        // chord length is (near-)zero.
```
Maybe the CPU shaders should assert that the chord length is less than some epsilon (I'm not sure if Rust panics on a floating point division by zero).

raphlinus replied:
It does not. I'll add an assert and some comments, that should help us debug the case.

And fun! When I added the assertion it got triggered. I'll add the robustness logic for this now.

armansito commented on src/cpu_shader/euler.rs:
```diff
@@ -0,0 +1,277 @@
+            // Rationale: this happens when fitting a cusp or near-cusp with
+            // a near 180 degree u-turn. The actual ES is bounded in that case.
+            // Further subdivision won't reduce the angles if actually a cusp.
+            return 2.0;
```
As we discussed briefly, I would elaborate on the choice of `2.0` here: what is the expected downstream effect and why is that acceptable.

armansito replied:
Maybe add a short sentence about the effect of `2.0` vs another number like `1.0`, e.g. "Returning an error value of 2.0 terminates the subdivisions because so and so".

armansito commented on src/cpu_shader/euler.rs:
```diff
@@ -0,0 +1,277 @@
+            return 2.0;
+        }
+        let e0 = (2. / 3.) / (1.0 + cth0);
+        let e1 = (2. / 3.) / (1.0 + cth1);
```
Is there are scenario in which `cth0` and `cth1` are both negative (so they pass the check above) and at least one of them has a value in the vicinity of -1?

armansito replied:
Consider adding assertions here that `cth0` and `cth1` won't cause divisions by zero.

raphlinus replied:
Ok, so when I add those assertions they do trigger on tricky_strokes, specifically 'M0 0.5 C 0 0 20 0 10 0`. I'm going to max with 1e-9 here (and in wgsl) to protect against divide by zero, as the correct behavior is to subdivide.

armansito commented on src/cpu_shader/euler.rs:
```diff
@@ -0,0 +1,277 @@
+/// Integrate Euler spiral.
+///
+/// TODO: investigate needed accuracy. We might be able to get away
+/// with 8th order.
```
Eventually we should have a wiki page (as well as the definitive paper) that explain the formulations here. Maybe add a disclaimer for now that a lot of the magic numbers here were derived by hand and needs to be documented?

raphlinus replied:
I wrote a paragraph explaining the tradeoff. It's also something that can potentially be explained in the paper.

armansito commented on src/cpu_shader/flatten.rs:
```diff
@@ -202,17 +149,68 @@ fn output_two_lines_with_transform(
+) {
+    let mut p0 = transform.apply(begin);
+    let mut r = begin - center;
+    let tol: f32 = 0.5;
```
Is `0.5` the tolerance value we want to go with?

raphlinus replied:
That's a very good question and we should take a closer look. One other thing I've noticed in testing is the second longpathdash now exceeds memory limits, and we should really understand why.

armansito commented on src/cpu_shader/flatten.rs:
```diff
@@ -202,17 +149,68 @@ fn output_two_lines_with_transform(
+
+    let th = angle / (n_lines as f32);
+    let c = th.cos();
+    let s = th.sin();
```
nit: This could be rewritten as `let (s, c) = th.sin_cos();`

armansito commented on src/cpu_shader/flatten.rs:
```diff
@@ -202,17 +149,68 @@ fn output_two_lines_with_transform(
     lines: &mut [LineSoup],
     bbox: &mut IntBbox,
 ) {
+    log!("@@@ flatten_euler: {:#?}", cubic);
```
nit: comment out or remove debug log?

armansito commented on src/cpu_shader/flatten.rs:
```diff
@@ -202,17 +149,68 @@ fn output_two_lines_with_transform(
+        MAX_LINES
+    } else {
+        MAX_LINES.min((std::f32::consts::TAU / theta).ceil() as u32)
+    };
```
This formulation is different from what's in the WGSL (and the removed lines below):

```
let MIN_THETA = 0.0001;
let tol = 0.1;
let radius = max(tol, length(p0 - transform_apply(transform, center)));
let theta = max(MIN_THETA, 2. * acos(1. - tol / radius));

// Always output at least one line so that we always draw the chord.
let n_lines = max(1u, u32(ceil(angle / theta)));
```

I believe the WGSL version is correct (apart from the tolerance and MIN_THETA values, which need tuning). I would keep the CPU and WGSL versions consistent.

raphlinus replied:
I hadn't intended to change this. I'm pretty sure that what happened is that #415 didn't make it into the euler-spiral-strokes branch. I've restored it, and good catch.

armansito commented on shader/flatten.wgsl:
```diff
@@ -188,96 +357,118 @@ fn flatten_cubic(cubic: CubicPoints, path_ix: u32, local_to_device: Transform, o
-        keep_params[i] = params;
-        val += params.val;
-        qp0 = qp2;
+    // Drop zero length lines.
```
Maybe mention here why the exact matches are valid (e.g. any invariants about prior detection of zero-length lines).

armansito commented on shader/flatten.wgsl:
```diff
@@ -188,96 +357,118 @@ fn flatten_cubic(cubic: CubicPoints, path_ix: u32, local_to_device: Transform, o
+        if t0 == 1.0 {
+            break;
+        }
+        loop {
```
Should there be a short explainer here about the derivative threshold, the iterative search to handle cusps, and the 3 cases below?

raphlinus replied:
That feels more like paper than code, but I'll add a sentence each.

armansito replied:
I agree. I think the paper will be the in-depth reference for all of this BUT I would like the code to eventually contain explainers and relevant citations as much as possible. I think the comments can be improved but I'm OK with the current state of things.

armansito commented on src/cpu_shader/flatten.rs:
```diff
@@ -1,94 +1,51 @@
-    val: f32,
-    a0: f32,
-    a2: f32,
+// TODO: remove this
```
Remove uses of `log!`? Or should we leave them in as a no-op?

P.S.: IWBN to have a debug macro that's optionally enabled for this purpose in general.

raphlinus replied:
I'm ok leaving them in for the time being, I found it useful while developing to be able to turn it on. I agree that a more sophisticated log story would be nice.

armansito commented on src/scene.rs:
```diff
@@ -116,7 +116,7 @@ impl Scene {
         const STROKE_TOLERANCE: f64 = SHAPE_TOLERANCE;
 
-        const GPU_STROKES: bool = false; // Set this to `true` to enable GPU-side stroking
+        const GPU_STROKES: bool = true; // Set this to `true` to enable GPU-side stroking
```
🎉

armansito reviewed (approved):
Great progress and exciting milestone! I left a number of minor suggestions, largely around documentation. This will get better as we work through the writeup soon but I think some things could be explained briefly in the code now.

raphlinus commented:
Ok, I think I've addressed the review comments, but I may have missed one.

Remaining issues:

* investigate why longpathdash is taking more memory
* revisit tolerance tuning parameters

These two items may not be completely separate :)
